### ISSUE TITLE
Use the native `<dialog>` element for Popovers/modals

### ISSUE THREAD

**Issue by aristath** [WordPress/gutenberg#51115]
## What problem does this address?
Our current popovers are custom implementations. We use `<div>` elements for everything, and have a lot of custom JS to handle the accessibility of those: Handling <kbd>esc</kbd> events, focus, keyboard navigation and so on.
There is however an alternative...
The native HTML [`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog) element handles most of these things automatically and is semantically the right element to use for modals, popovers & dialogs. It is stylable, supported by all browsers, and will allow us to clean up a lot of code while at the same time improving the project's a11y.

**Comment by ciampo** [comment]
Hey @aristath , thank you for opening the issue! 

The `<dialog />` element won't probably adapt to all the needs that we have for a popover — for example, a tooltip is not a dialog.

A new [Popover API](https://caniuse.com/mdn-api_htmlelement_popover) is starting to land in browsers, but it's still early days (also, [see the difference between popover and dialog APIs](https://developer.chrome.com/blog/introducing-popover-api/#the-difference-between-a-popover-and-a-dialog)).

Finally, even in a scenario where these APIs would be stable and well-supported by browsers, it may not be straightforward for our `@wordpress/components` package to use them, since we have a series of components that all rely on `Popover`, and some of them have complex implementations (like DropdownMenu).

We're also migrating such components to third-party headless libraries that comply with accessibility requirements while offering a refined user experience.

**Comment by aristath** [comment]
> The `<dialog />` element won't probably adapt to all the needs that we have for a popover — for example, a tooltip is not a dialog.

Then perhaps a tooltip should use a different component?
We have many scenarios for which the `<dialog>` element would be a perfect fit... So perhaps we could introduce a new component and migrate those scenarios to the new one?

**Comment by afercia** [comment]
Just adding some technical details:

So far, the `<dialog>` element couldn't be used because its initial focus behavior was not spec'ed. On 26 January 2023 (that is 4 months ago), the HTML working group finally agreed on the initial focus behavior after a mylti-year discussion. That's now detailed [in the specification](https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element). However, that's doesn't mean the initial focus behavior is actually fully implemented in the browsers we support.
Also, I'm not sure there's any data yet about actual [screen readers support](https://a11ysupport.io/tech/html/dialog_element).
For these reasons, we may want to wait a bit before considering using the `<dialog>` element.

More details:

[Dialog Focus in Screen Readers - (Spec) Update: 26 January 2023](https://adrianroselli.com/2020/10/dialog-focus-in-screen-readers.html#Update01)
by Adrian Roselli

[Use the dialog element (reasonably)](https://www.scottohara.me/blog/2023/01/26/use-the-dialog-element.html)
by Scott O’Hara - 2023.01.26

WebKit tracking bug:
https://bugs.webkit.org/show_bug.cgi?id=84635

Firefox tracking bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=840640

**Comment by ciampo** [comment]
As of today, I don't see us needing to use `<dialog />` or the new popover APIs — at least in `@wordpress/components`.

We can evaluate in some time, after those technologies have matured and received wider adoption — but given that we're slowly migrating to use third-party libraries for our components (including the ones using popovers), I personally think that it won't likely make much sense (unless there's a clear improvement that can justify the effort).

**Comment by manzoorwanijk** [comment]
[`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog) is now considered widely available and the [`popover` API](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/popover) is also available in all the latest browsers since April 2024.

**Comment by mirka** [comment]
Closing for the [reasons](https://github.com/WordPress/gutenberg/issues/51115#issuecomment-1570472450) raised by @ciampo.

**Comment by afercia** [comment]
>  I personally think that it won't likely make much sense (unless there's a clear improvement that can justify the effort).

Reopening for broader consideration. This needs to be considered from an accessibility perspective too. Simply put: when there is a native HTML feature, use that instead of replicating with JS. 

> given that we're slowly migrating to use third-party libraries for our components 

Re: third-party libraries, they shoud be evaluated also from this perspective before their adoption.

Cc @joedolson FYI

**Comment by ciampo** [comment]
**Specifically for the WordPress Design System, ie. UI componentry used to build Gutenberg and related**

Using third-party libraries is a deliberate decision that lets us delegate the choice of lower-level primitives and functionality, trusting that such libraries will provide the most accessible and feature-rich UX for our users.

**Whether our implementations should use a native HTML `<dialog />` is therefore a decision that should be made (in our case) at the base UI level. And it looks like it has already been [evaluated and discarded](https://github.com/mui/base-ui/issues/485#issuecomment-2241703553) for now.**

Also note that other large libraries (such as React Aria) [came to a similar conclusion](https://github.com/adobe/react-spectrum/discussions/9696#discussioncomment-15942257): the `dialog` element works well for simple / standard cases, but it doesn't yet cover more advanced usecase. 

I'm certain that it will be adopted in all major libraries at some point, but that time hasn't come yet. And it's not even our decision to make, at this point.

---

If we're referring to how blocks should implement their front-end overlays, that's a separate topic where I don't have a particularly strong opinion, and maybe the HTML `dialog` element (and popover APIs) are a good fit.

**Comment by afercia** [comment]
Thank you for your feedback. Those are all points that need to be taken into consideration.

> Using third-party libraries is a deliberate decision ...

Can you point me at a public discussion on this project where that decision has been made? I wonder how such decisions can be made without involving other teams. Especially this decision, which deeply impacts accessibility, should have been discussed with the accessibility team. In case there hasn't been any broader discussion about it, I'm not sure that's the best way to encourage collaboration on this project. 

> And it looks like it has already been https://github.com/mui/base-ui/issues/485#issuecomment-2241703553 for now.

The post you linked to is from Jul 21, 2024, which is more that two years ago. I do see it references at least one feature of the `<dialog>` element that has changed in the meantime in the native implementation so I would consider that post outdated.

**Comment by ciampo** [comment]
> Can you point me at a public discussion on this project where that decision has been made? I wonder how such decisions can be made without involving other teams. Especially this decision, which deeply impacts accessibility, should have been discussed with the accessibility team. In case there hasn't been any broader discussion about it, I'm not sure that's the best way to encourage collaboration on this project.

No decision was made recently. UI componentry in Gutenberg has been relying on third-party headless component libraries for at least 6+ years, for sure since Diego Haz (who worked on the Gutenberg project at the time) started his `reakit` project (later renamed to `ariakit`), which has been a dependency since then.

The @WordPress/gutenberg-components team has been slowly and progressively increasing the reliance on reakit/ariakit (and base ui since last year), with plenty of collaboration with other teams, including the accessibility team. 

> I'm not sure that's the best way to encourage collaboration on this project.

Same reply as https://github.com/WordPress/gutenberg/issues/81634#issuecomment-5343009683 applies:

_**Accessibility is a shared responsibility, rather than a sign-off owned by one team.** @WordPress/gutenberg-components and @WordPress/gutenberg-design contributors should be trusted to consider it in our normal work and to seek specialist input when specific risks or uncertainty warrant it._

> The post you linked to is from Jul 21, 2024, which is more that two years ago. I do see it references at least one feature of the `<dialog>` element that has changed in the meantime in the native implementation so I would consider that post outdated.

Let's not nitpick based on a single link that I sent. The thread continues, and there are several recent replies, including https://github.com/mui/base-ui/issues/485#issuecomment-4014854466. Pretty much every state-of-the-art UI component library is still not using `dialog` for the same reasons. I'm sure that when the technology is ripe enough, they will switch.

**Comment by mirka** [comment]
Let's revisit when native `dialog` is ready, and will actually provide benefits over the current library implementations.

**Comment by westonruter** [comment]
Per @manzoorwanijk in https://github.com/WordPress/gutenberg/issues/51115#issuecomment-3135814732:

> [`<dialog>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/dialog) is now considered widely available and the [`popover` API](https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Global_attributes/popover) is also available in all the latest browsers since April 2024.

That said, I think the primary benefit for `DIALOG` would be to implement things like the mobile nav menu paired with Invoker Commands. This will allow for interactivity without any JavaScript. See https://github.com/WordPress/gutenberg/pull/69720.
